torch/profiler/_utils.py

```diff
@@ -194,6 +203,86 @@ def compute_idle_time(self):
             self.metrics[EventKey(event)].idle_time_ns = EventKey(
                 event).intervals_overlap(idle_intervals)
 
+    def rank_events(self, length):
+        '''
+        Filter and Rank the events based on some heuristics:
+        1) Events that are in the falling phase of the queue depth.
+        2) Events that have a high idle_time, self_time difference.
+
+        Parameters:
+            length: The number of events to return.
+        '''
+
+        # Find the interval when qd is falling to 0
+        queue_depth_list = list(reversed(self.queue_depth_list))
+        qd_values = [e.queue_depth for e in queue_depth_list]
+
+        decrease_interval = []
+        for i in range(len(qd_values) - 1):
```

Comment:
This is a very deep nested structure and is hard to follow. That's generally a sign that some of the inner blocks should be factored into named functions.